### ISSUE TITLE
Behaviour change in session handling in 9.4.21.v20190926

### ISSUE THREAD

**Issue by dreis2211** [eclipse/jetty.project#4158]
Hi,

I've been trying for the last couple of days to replicate a session related issue in one of Spring-Boot's tests reported at https://github.com/spring-projects/spring-boot/issues/18494. The mentioned test there fails because - unlike before - a new session is created every time via `((HttpServletRequest) request).getSession(true)` and thus breaks the logic of persisting attributes.

I finally could write a test in Jetty that replicates the behaviour, which passes on `9.4.20.v20190813` and fails on `9.4.21.v20190926`:

```
public class CookieRequestSessionTest
{
	private Server server;
	private HttpClient client;

	@BeforeEach
	public void startServer() throws Exception
	{
		server = new Server();
		ServerConnector connector = new ServerConnector(server);
		connector.setPort(0);
		server.addConnector(connector);

		// Default session behavior
		ServletContextHandler contextHandler = new ServletContextHandler(ServletContextHandler.SESSIONS);
		contextHandler.setContextPath("/");
		contextHandler.addServlet(SessionServlet.class, "/session");
		contextHandler.addServlet(DefaultServlet.class, "/");

		HandlerList handlers = new HandlerList();
		handlers.addHandler(contextHandler);
		handlers.addHandler(new DefaultHandler());

		server.setHandler(handlers);

		server.start();
	}

	@AfterEach
	public void stopServerAndClient()
	{
		LifeCycle.stop(server);
		LifeCycle.stop(client);
	}

	@BeforeEach
	public void startClient() throws Exception
	{
		client = new HttpClient();
		client.start();
	}

	@Test
	public void testWithCookieInRequest() throws Exception
	{
		ContentResponse response = client.newRequest(server.getURI().resolve("/session"))
				.method(HttpMethod.POST)
				.header("Cookie", "JSESSIONID=123")
				.send();
		assertThat("Response status", response.getStatus(), is(HttpStatus.OK_200));
		String firstSessionId = response.getContentAsString();

		response = client.newRequest(server.getURI().resolve("/session"))
				.method(HttpMethod.POST)
				.header("Cookie", "JSESSIONID=123")
				.send();
		assertThat("Response status", response.getStatus(), is(HttpStatus.OK_200));
		String secondSessionId = response.getContentAsString();

		assertEquals(firstSessionId, secondSessionId);
	}

	public static class SessionServlet extends GenericServlet
	{
		@Override
		public void service(ServletRequest request, ServletResponse response) throws IOException {
			HttpSession session = ((HttpServletRequest) request).getSession(true);
			PrintWriter writer = response.getWriter();
			writer.append(session.getId());
		}
	}
}
```

This appears to be connected to the usage of a cookie with a supplied JSESSIONID. When I comment out the lines that add the cookie header the test passes. I suspect this is a bug, but I'm not sure. Any help is highly appreciated.

Cheers,
Christoph

**Comment by gregw** [comment]
The issue is that the second request sent by the test has two JSESSIONID cookies - the first being the 123 set explicitly and the second being the cookie set by the HttpClient from the Set-Cookie in the response of the first request.

Previously, when given multiple cookies, we checked them all for a valid session.   With 9.4.21, we check the first, see it is invalid and return null - so a new session is then created.

So I'm not entirely sure that it is valid to send duplicate cookies like that, so results are going to be unpredictable... we should at least work out if this was a deliberate change or an side effect of other changes @janbartel ?

**Comment by joakime** [comment]
@dreis2211 to see the detail of what @gregw pointed out you can make the following adjustments to your test case ...

``` java
package jetty;

import java.io.IOException;
import java.io.PrintWriter;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;

import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.client.api.ContentResponse;
import org.eclipse.jetty.client.api.Request;
import org.eclipse.jetty.client.api.Response;
import org.eclipse.jetty.http.HttpMethod;
import org.eclipse.jetty.http.HttpStatus;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.ServerConnector;
import org.eclipse.jetty.server.handler.DefaultHandler;
import org.eclipse.jetty.server.handler.HandlerList;
import org.eclipse.jetty.servlet.DefaultServlet;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.util.component.LifeCycle;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class CookieRequestSessionTest
{
    private Server server;
    private HttpClient client;

    @BeforeEach
    public void startServer() throws Exception
    {
        server = new Server();
        ServerConnector connector = new ServerConnector(server);
        connector.setPort(0);
        server.addConnector(connector);

        // Default session behavior
        ServletContextHandler contextHandler = new ServletContextHandler(ServletContextHandler.SESSIONS);
        contextHandler.setContextPath("/");
        contextHandler.addServlet(SessionServlet.class, "/session");
        contextHandler.addServlet(DefaultServlet.class, "/");

        HandlerList handlers = new HandlerList();
        handlers.addHandler(contextHandler);
        handlers.addHandler(new DefaultHandler());

        server.setHandler(handlers);

        server.start();
    }

    @AfterEach
    public void stopServerAndClient()
    {
        LifeCycle.stop(server);
        LifeCycle.stop(client);
    }

    @BeforeEach
    public void startClient() throws Exception
    {
        client = new HttpClient();
        client.start();
    }

    @Test
    public void testWithCookieInRequest() throws Exception
    {
        ContentResponse response = client.newRequest(server.getURI().resolve("/session"))
            .method(HttpMethod.POST)
            .header("Cookie", "JSESSIONID=123")
            .send();
        dumpExchange(response);
        assertThat("Response status", response.getStatus(), is(HttpStatus.OK_200));
        String firstSessionId = response.getContentAsString();

        response = client.newRequest(server.getURI().resolve("/session"))
            .method(HttpMethod.POST)
            .header("Cookie", "JSESSIONID=123")
            .send();
        dumpExchange(response);
        assertThat("Response status", response.getStatus(), is(HttpStatus.OK_200));
        String secondSessionId = response.getContentAsString();

        assertEquals(firstSessionId, secondSessionId);
    }

    private void dumpExchange(Response response)
    {
        Request request = response.getRequest();

        System.out.println();
        System.out.println("-- REQUEST --");
        System.out.printf("%s %s %s%n", request.getMethod(), request.getURI(), request.getVersion());
        System.out.println(request.getHeaders());
        System.out.println("-- RESPONSE --");
        System.out.printf("%s %d %s%n", response.getVersion(), response.getStatus(), response.getReason());
        System.out.println(response.getHeaders());
    }

    public static class SessionServlet extends HttpServlet
    {
        @Override
        protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException
        {
            HttpSession session = request.getSession(true);
            PrintWriter writer = response.getWriter();
            writer.append(session.getId());
        }
    }
}
```

Which shows the following output ...

``` 
-- REQUEST --
POST http://127.0.1.1:39325/session HTTP/1.1
Accept-Encoding: gzip
User-Agent: Jetty/9.4.21.v20190926
Cookie: JSESSIONID=123
Host: 127.0.1.1:39325


-- RESPONSE --
HTTP/1.1 200 OK
Date: Mon, 07 Oct 2019 14:48:18 GMT
Set-Cookie: JSESSIONID=node01qsjgdf88ygpk1vr13ovv7649d0.node0; Path=/
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Content-Length: 32
Server: Jetty(9.4.21.v20190926)



-- REQUEST --
POST http://127.0.1.1:39325/session HTTP/1.1
Accept-Encoding: gzip
User-Agent: Jetty/9.4.21.v20190926
Cookie: JSESSIONID=123
Host: 127.0.1.1:39325
Cookie: JSESSIONID=node01qsjgdf88ygpk1vr13ovv7649d0.node0


-- RESPONSE --
HTTP/1.1 200 OK
Date: Mon, 07 Oct 2019 14:48:18 GMT
Set-Cookie: JSESSIONID=node01x6vsx618e9uzwnys07r2h9qa1.node0; Path=/
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Content-Length: 31
Server: Jetty(9.4.21.v20190926)


```

**Comment by dreis2211** [comment]
Thanks @joakime , but I already did something similar. I also played around with the code already to bring back the old behaviour just out of interest:
```patch
--- a/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java
@@ -1633,7 +1633,11 @@ public class SessionHandler extends ScopedHandler
                             LOG.debug("Got Session ID {} from cookie {}", requestedSessionId, sessionCookie);
                         if (requestedSessionId != null)
                         {
-                            break;
+                            session = getHttpSession(requestedSessionId);
+                            if (session != null && isValid(session))
+                            {
+                                break;
+                            }
                         }
                     }
                 }
```
Let's see what @janbartel has to say.
Cheers.

**Comment by janbartel** [comment]
According to the Cookie specification, the user-agent should never send duplicate cookies with the same name, domain, path etc. Similarly, the user-agent should never send a cookie which it did not receive in a Set-Cookie header on a response.

RFC6265:
Section 3: 

> To store state, the origin server includes a Set-Cookie header in an
>  HTTP response.  In subsequent requests, the user agent returns a
>  Cookie request header to the origin server.  The Cookie header
>  contains cookies the user agent received in previous Set-Cookie
>  headers.

Section 5.3:
User agent processing of Set-Cookie header:
>  11.  If the cookie store contains a cookie with the same name,
>         domain, and path as the newly created cookie:
> 
>         1.  Let old-cookie be the existing cookie with the same name,
>             domain, and path as the newly created cookie.  (Notice that
>             this algorithm maintains the invariant that there is at most
>             one such cookie.)
> 
>         2.  If the newly created cookie was received from a "non-HTTP"
>             API and the old-cookie's http-only-flag is set, abort these
>             steps and ignore the newly created cookie entirely.
> 
>         3.  Update the creation-time of the newly created cookie to
>             match the creation-time of the old-cookie.
> 
>         4.  Remove the old-cookie from the cookie store.

Moreover, according to the spec, the user-agent should never generate multiple Cookie headers:
Section 5.4:
>    When the user agent generates an HTTP request, the user agent MUST
>    NOT attach more than one Cookie header field.

So taken all together, it seems that a server should never receive multiple Cookie headers, nor multiple cookies with the same name & domain & path & expiry.

**Comment by gregw** [comment]
@janbartel However I think it is actually possible to have overlapping domains.
A Cookie can be set for both '/' and '/foo', which are different domains, so a request to '/foo/bar' will have both cookies set on it.   Once upon a time the intention was to also send the domain with each cookie, but that spec died and now we just get the duplicate cookies.   

Regardless of the legality, it is apparent that at least curl and some browsers are capable of sending duplicate cookies.   This is reflected by the fact  that we do have a loop to look for session cookies.
So just looking at the first cookie doesn't really feel like the right thing to do!  We should either error on duplicate cookies or look at them all.  Perhaps we look at them all, check for validity and only if we get two valid sessions do we do a 400 Bad Request?

**Comment by janbartel** [comment]
I'm happy to re-instate the previous behaviour of looking through duplicate session cookies until a valid one is found. It would be new behaviour to exhaustively examine them and throw a 400 Bad Request if more than one is valid.

**Comment by janbartel** [comment]
PR #4168

**Comment by janbartel** [comment]
Fix commited with merge of PR #4168
